### DISCUSSION TITLE
SyntaxError: decimal numbers not supported

### DISCUSSION THREAD

**Discussion by synapticonashaikh** [micropython/micropython#9896] [Core Development] [answered]
Hello, I'm working on a project in which I'm using the sample code of the micropython minimal project (I'm using the STM32F446xx board).
I've redefined the standard libraries to be able to work for my board. Currently, I can generate the minimal binary file (it is working fine). As the sample project is bare minimal, it does not support the floating point operation. Hence, I can't use any floating point operations. I would like to know how I could use the floating point operation. If anyone has done that before, please help me with some guidance.

**Comment by dlech** [comment, marked as answer]
You will need to change to `MICROPY_FLOAT_IMPL`.

    #define MICROPY_FLOAT_IMPL                      (MICROPY_FLOAT_IMPL_FLOAT)

And make sure ` -mfpu=fpv4-sp-d16 -mfloat-abi=hard` is added to `CFLAGS` somewhere in the Makefile.

**Reply by synapticonashaikh** [reply]
Hello, after adding `-mfpu=fpv4-sp-d16 -mfloat-abi=hard` to ` CFLAGS` I got the following error: 
`arm-none-eabi-gcc: error: -mfloat-abi=soft and -mfloat-abi=hard may not be used together`

**Reply by robert-hh** [reply]
In that case remove -mfloat-abi=soft

**Reply by robert-hh** [reply]
Or try `-mfloat-abi=softfp` instead of both `-mfloat-abi=soft` and `-mfloat-abi=hard`.

**Reply by synapticonashaikh** [reply]
1. -`mfloat-abi=soft` is not present in the makefile. I'm not sure from where it is getting the definition. 
2. I tried adding `-mfloat-abi=softfp ` but then it is giving lots of undefined errors in the formatfloat file. Some errors are as follows:
```
arm-none-eabi-ld: build/py/mpprint.o: in function `mp_vprintf':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/mpprint.c:546: undefined reference to `__aeabi_d2f'
arm-none-eabi-ld: build/py/formatfloat.o: in function `mp_format_float':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:118: undefined reference to `__aeabi_fcmpun'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:133: undefined reference to `__aeabi_fcmpun'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:133: undefined reference to `__aeabi_fcmple'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:138: undefined reference to `__aeabi_fcmpun'
```

**Reply by synapticonashaikh** [reply]
I'm attaching the make file! I didn't modify the file that much. I only added a few of my files (with peripheral init functionality) along with the main.c. reference https://github.com/micropython/micropython/blob/master/ports/minimal/Makefile
[Makefile.txt](https://github.com/micropython/micropython/files/9973018/Makefile.txt)

**Reply by dlech** [reply]
> 1. -`mfloat-abi=soft` is not present in the makefile. I'm not sure from where it is getting the definition.

It is on this line as `-msoft-float `: https://github.com/micropython/micropython/blob/eefd946e60aba3ac61c7bfbd0272d07be289e3f3/ports/minimal/Makefile#L25

Delete that and replace it with `-mfpu=fpv4-sp-d16 -mfloat-abi=hard`. Or don't change it and continue to use soft float support. But since the chip has hardware support for floats, I don't see a reason to use soft float.

**Reply by synapticonashaikh** [reply]
I figured that `-msoft-float` might be the `-mfloat-abi=soft`. So, I removed that and tried with both `-mfpu=fpv4-sp-d16 -mfloat-abi=hard` and `-mfloat-abi=softfp`, but I'm still getting lots of undefined errors (same as the previous one).

**Reply by dlech** [reply]
Did you do `make clean` after changing the Makefile?

**Reply by synapticonashaikh** [reply]
> Did you do `make clean` after changing the Makefile?
Yes, after every try and variation, I'm doing a make clean.

**Reply by dlech** [reply]
Maybe need to link libgcc?

    LIBS += "$(shell $(CC) $(CFLAGS) -print-libgcc-file-name)"

**Reply by synapticonashaikh** [reply]
I replaced `LIBS =` with `LIBS += "$(shell $(CC) $(CFLAGS) -print-libgcc-file-name)"` still I'm getting these errors:
```
arm-none-eabi-ld: build/py/formatfloat.o: in function `mp_format_float':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:187: undefined reference to `powf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:190: undefined reference to `powf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:238: undefined reference to `powf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:241: undefined reference to `powf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/formatfloat.c:313: undefined reference to `powf'
arm-none-eabi-ld: build/py/parsenum.o:/home/synapticon/Desktop/micropython/ports/minimal/../../py/parsenum.c:354: more undefined references to `powf' follow
arm-none-eabi-ld: build/py/objcomplex.o: in function `complex_unary_op':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/objcomplex.c:130: undefined reference to `sqrtf'
arm-none-eabi-ld: build/py/objcomplex.o: in function `mp_obj_complex_binary_op':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/objcomplex.c:234: undefined reference to `sqrtf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objcomplex.c:242: undefined reference to `logf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objcomplex.c:243: undefined reference to `atan2f'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objcomplex.c:246: undefined reference to `expf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objcomplex.c:247: undefined reference to `cosf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objcomplex.c:248: undefined reference to `sinf'
arm-none-eabi-ld: build/py/objfloat.o: in function `mp_obj_float_divmod':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/objfloat.c:216: undefined reference to `fmodf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objfloat.c:235: undefined reference to `floorf'
arm-none-eabi-ld: build/py/objfloat.o: in function `mp_obj_float_binary_op':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/objfloat.c:288: undefined reference to `fmodf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objfloat.c:303: undefined reference to `floorf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/objfloat.c:316: undefined reference to `powf'
arm-none-eabi-ld: build/py/modbuiltins.o: in function `mp_builtin_round':
/home/synapticon/Desktop/micropython/ports/minimal/../../py/modbuiltins.c:512: undefined reference to `powf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/modbuiltins.c:514: undefined reference to `nearbyintf'
arm-none-eabi-ld: /home/synapticon/Desktop/micropython/ports/minimal/../../py/modbuiltins.c:517: undefined reference to `nearbyintf'
make: *** [Makefile:86: build/firmware.elf] Error 1
```

**Reply by dlech** [reply]
Now you need to add the `libm` source files to the build.

```make
SRC_LIB += $(addprefix lib/,\
        libm/math.c \
        libm/fmodf.c \
        libm/nearbyintf.c \
        libm/ef_sqrt.c \
        libm/kf_rem_pio2.c \
        libm/kf_sin.c \
        libm/kf_cos.c \
        libm/kf_tan.c \
        libm/ef_rem_pio2.c \
        libm/sf_sin.c \
        libm/sf_cos.c \
        libm/sf_tan.c \
        libm/sf_frexp.c \
        libm/sf_modf.c \
        libm/sf_ldexp.c \
        libm/asinfacosf.c \
        libm/acoshf.c \
        libm/asinhf.c \
        libm/atanf.c \
        libm/atan2f.c \
        libm/atanhf.c \
        libm/log1pf.c \
        libm/sf_erf.c \
        libm/erf_lgamma.c \
        libm/wf_lgamma.c \
        libm/wf_tgamma.c \
        libm/roundf.c \
	)
```

**Reply by synapticonashaikh** [reply]
Hello, I added the `libm` source files to the build (as you mentioned).
And then I add the object of it to the final build! 
```
LIBM_O = $(addprefix $(BUILD)/, $(SRC_LIB:.c=.o))
OBJ += $(LIBM_O)
```
Now the code is compiling without any errors! But after every equation, the interpreter is getting hanged/frozen. It is getting hanged on all commands (even simple `print("hello")`).

[Makefile.txt](https://github.com/micropython/micropython/files/9979412/Makefile.txt)
![image](https://user-images.githubusercontent.com/115145853/201054265-386ae989-68e3-43b8-b40a-d619ca1def21.png)

**Reply by dlech** [reply]
Does "hello world" start working again if you disable `MICROPY_FLOAT_IMPL`?

**Reply by synapticonashaikh** [reply]
I commented the `#CFLAGS += -DMICROPY_FLOAT_IMPL=MICROPY_FLOAT_IMPL_FLOAT`, but still, the interpreter is getting frozen (after entering any command) like earlier.

**Reply by synapticonashaikh** [reply]
Hello, thanks for all the support! I'm able to perform the floating point operation now! I did the step-by-step debugging and found that somehow `-mfloat-abi=hard` is creating an issue. I replaced it with the previous definition, `-mfloat-abi=solft`, and it is working now.
![image](https://user-images.githubusercontent.com/115145853/201351705-b48cf3da-9bda-4f0e-a030-433014471c23.png)
[Makefile.txt](https://github.com/micropython/micropython/files/9990557/Makefile.txt)
